nakadi/event_stream.py

```diff
@@ -7,72 +7,94 @@
 import sys
 
 
-def create_stream_generator(kafka_pool, topic, partition, start_from, opts):
+def create_stream_generator(kafka_pool, topic, cursors, opts):
+
+    topics = {}
+    for cursor in cursors:
+        topic_partition = (topic, int(cursor['partition']))
+        topics[topic_partition] = int(cursor['offset'])
+
+    partitions = [int(cursor['partition']) for cursor in cursors]
 
     start = datetime.datetime.now()
 
     # stream
     def generator():
-        messages_read = 0
+
         keep_alive_in_a_row = 0
-        current_batch = []
-        batch_start_time = datetime.datetime.now()
 
+        messages_read = 0
+        current_batch = dict(zip(partitions, [[] for _ in partitions]))
+
+        batch_start_time = datetime.datetime.now()
         with kafka_pool.kafka_client() as client:
 
-            consumer = KafkaConsumer((topic, partition, start_from),
-                         kafka_client = client,
-                         auto_commit_enable = False,
-                         consumer_timeout_ms = 200)
+            consumer = KafkaConsumer(topics,
+                                     kafka_client=client,
+                                     auto_commit_enable=False,
+                                     consumer_timeout_ms=200)
 
             while True:
                 try:
                     message = consumer.next()
-
                     # if we read the message - reset keep alive counter
                     keep_alive_in_a_row = 0
 
                     # put message to batch
                     messages_read += 1
                     message_json = json.loads(message.value.decode('utf-8'))
-                    current_batch.append(message_json)
+                    current_batch[message.partition].append(message_json)
 
                 except ConsumerTimeout:
                     pass
 
                 # check if it's time to send the batch
                 time_since_batch_start = datetime.datetime.now() - batch_start_time
-                if time_since_batch_start.total_seconds() >= opts['batch_flush_timeout'] != 0 or len(current_batch) >= opts['batch_limit']:
-
-                    # send the messages we could read so far
-                    if len(current_batch) > 0:
-                        stream_message = __create_stream_message(partition, start_from + messages_read, current_batch)
-                        with __measure_time(current_batch, stream_message):
-                            yield stream_message
-
-                    # just send the keep alive
-                    else:
-                        # if we hit keep alive count limit - close the stream
+                batch_sum = sum([len(batch_per_partition) for batch_per_partition in current_batch.values()])
+                latest_offsets = consumer.offsets("fetch")
+
+                if time_since_batch_start.total_seconds() >= opts['batch_flush_timeout'] != 0 or batch_sum >= opts['batch_limit']:
```

Comment:
Yes, `connexion` takes care or that types based on the Swagger definition of the options type. Did not know about that when writing the comment :D